### ISSUE TITLE
Create a dynamic package version based on latest git tag

### ISSUE THREAD

**Issue by quiandy** [conan-io/conan#4449]
I'm still trying to craft a robust way to create conan packages as part of a process following a tag operation on our git repository.

The ideal operations would be -> merge2master -> create git tag -> create package for shipping
In our configuration, the conan recipe file is in the git repository with the code, so our ci system clones the repository and should take things from there

The idea is to have the conan package version match the tag version, and to be working on the latest tag which was created. To do it, we've resorted to injecting some custom code which runs upon the Conan class creation, and tries to implement some logic to get the latest tag.

Something like this:
```
class myConan():
	url = "[...]/myrepo.git"
	version, commitAtVersion  = get_version_from_latest_git_tag(url)
```
Now in get_version_from_latest_git_tag(url), we don't have access to the repository (unless we clone it), so I use a git-ls-remote command to reach out to the url and retrieve tag information (and commit information). git-ls-remote however can only retrieve limited info, e.g. I cannot sort the tags by tag creation date (and hence be sure I'm fetching the latest one).

Long story short, this looks clunky and brittle, what's the proper way to do something like this? Would the [get_tag() method recently added](https://github.com/conan-io/docs/pull/1020/commits/633de38fc0d0dff6858b11d7f41fae00e0e52629) to the trick? Can you provide some insight into how to use it in this case?

**Comment by Johnnyxy** [comment]
@quiandy the [conan-package-tools](https://github.com/conan-io/conan-package-tools) can be used to do this, check the docs [here](https://github.com/conan-io/conan-package-tools#package-version-based-on-commit-checksum).

Alternatively you can do it yourself by omitting the ```version``` class member and specify it at package-creation:

```python
class myConan():
    settings = ...
    options = ...
```

Specify the version string on command line (replace GIT_VERSION):
```
conan create your_package/GIT_VERSION@test/test -s ... -o ...
```

Best regards

**Comment by quiandy** [comment]
Thanks, but this just means that I have to move the package version deduction logic outside of the package recipe, so I can call the conan create with the version passed in.
Doesn't it make more sense to keep it inside the packaging framework rather then lying around?

**Comment by uilianries** [comment]
Hello @quiandy and everyone!

I see your problem, actually your idea is very common for some projects.

CPT (conan package tools) offers some helps to identify if your current branch is a tag and it branch name. As you want to create dynamic version, I would say to remove `version` from the recipe and set the reference on CPT.

You could take a in some examples:
- Flatbuffers: https://github.com/google/flatbuffers/blob/master/conan/build.py#L61
- uvw: https://github.com/skypjack/uvw/blob/master/.conan/build.py#L41

Also, CPT offers `ci_manager` module, a really nice helper, where you can check if your branch is a tag: https://github.com/conan-io/conan-package-tools/blob/develop/cpt/ci_manager.py#L83

Regards!

**Comment by quiandy** [comment]
@Johnnyxy : oh no, it is definitely possible to make the version dynamic by executing some logic inside the conan recipe. E.g. you could export some text file containing some information about the version, (using the '''exports''' keyword), and then have '''version=readVersion(somefile)''' set by a function which contains some parsing logic.

Now seeing the existance of CPT however,  this does seem a bit hacky. I'll definitely give CPT a deeper look. Thanks @Johnnyxy  and @uilianries!
